### DISCUSSION TITLE
[Feature Request] Debugger/Hardhat

### DISCUSSION THREAD

**Discussion by Vorh** [zkSync-Community-Hub/zksync-developers#17] [Feature requests]
### Network

zkSync Era

### Provide a brief description of the feature you would like to see implemented.

Hello! 

At the moment any fail transaction cannot debug.
This makes it difficult to develop, support and create new products at ZkSync. 

At the moment ZkSync not support standard dev tools (`hardhat`) which are used in the world to develop smart contracts because `ZkSync EVM` not compatible with `Ethereum ENV`

What should be done?

Add support [Hardhat Network](https://hardhat.org/hardhat-network/docs/overview) module for ZkSync. 
By default it module has backed by the [@ethereumjs/vm](https://www.npmjs.com/package/@ethereumjs/vm) EVM implementation, the same one used by ganache, Remix and Ethereum Studio.

For support ZkSync should be implement own version of `@ethereumjs/vm`

What will it give developers?

Allow use hardhat for local developing:
- allow fork node and run test on particular blockNumber
- allow emulate transaction on local node with debug
 
What will it give ZkSync?

increase in new protocols and products because development and testing is becomes the same as Ethereum ecosystem.

### Who does this feature request benefit the most (users, developers, protocol, etc.)?

developers

**Comment by uF4No** [comment]
Hi @Vorh 

Thanks for the feedback. Currently you can test applications locally using the zkSync local setup, which simulates L1/L2 locally and provides similar results as testnet/mainnet.

You can find more info about it in [this page from the docs](https://era.zksync.io/docs/tools/hardhat/testing.html).

We know there are some gaps in features with the Hardhat node but we're working on it.

**Reply by Vorh** [reply]
Sorry. How I can to debug fail transaction from ZkSync with local node? 
Did not see the information in the mentioned guide...

**Reply by bxpana** [reply]
You could use the logs from the local setup to see where an why a transaction is failing. Or at least get more info than you are getting now.

**Reply by Vorh** [reply]
How to emulate fail transaction from ZkSync on local node on particular blockNumber?
I read tutorial from link above but not find any information about it.
Do I can to view contract address or some other debug information where throw error?

**Reply by Vorh** [reply]
@uF4No @bxpana any news about support hardhat node?

**Reply by Vorh** [reply]
@uF4No @bxpana any news about support hardhat node?

**Reply by uF4No** [reply]
hey @Vorh the Hardhat node has some differences with the zkEVM on Era. We've just released [an alpha version of our local in-memory node](https://era.zksync.io/docs/tools/testing/era-test-node.html), which supports features like forking or replaying transactions from testnet/mainnet. It also supports using `console.log`  in your contracts.

You can find the [differences in features between the docker node and the in-memory node here](https://era.zksync.io/docs/tools/testing/).

The [in-memory node is open-source](https://github.com/matter-labs/era-test-node) so you can report any issues in the repo.

**Reply by Vorh** [reply]
@uF4No Thanks a lot! This is very good news!

**Comment by keytwojohn** [comment]
Integrating support for the Hardhat Network module into zkSync would greatly enhance the development experience for developers by allowing for local development and debugging using familiar tools. By implementing zkSync's own version of @ethereumjs/vm, developers can leverage features such as forking nodes and running tests on specific block numbers, as well as emulating transactions on a local node for debugging purposes. This improvement in developer tooling would facilitate faster iteration and testing, ultimately leading to increased innovation and adoption of zkSync protocols and products within the ecosystem.

**Comment by kimballthomas** [comment]
Adding support for Hardhat Network module in zkSync would indeed greatly benefit developers by enabling them to use familiar development tools for local development and testing. This enhancement would streamline the development process, allowing developers to debug failed transactions and test new features more effectively. Ultimately, it would lead to increased efficiency and innovation within the zkSync ecosystem.

**Comment by tovanpanet** [comment]
Implementing support for the Hardhat Network module in zkSync would greatly benefit developers by enabling them to use familiar development tools like Hardhat for local development and testing. This enhancement would streamline the development process and foster the creation of new protocols and products on zkSync.

**Comment by maestaspat** [comment]
Implementing support for Hardhat Network module in zkSync would greatly benefit developers by enabling local development, testing, and debugging of smart contracts, leading to faster and more efficient development cycles.

**Comment by sifrediblue** [comment]
The user is requesting the addition of support for the Hardhat Network module on the zkSync Era network to enable better debugging of failed transactions and to allow for local development using Hardhat. This feature request would benefit developers and would increase the development and testing of new protocols and products on the zkSync Era network. To support Hardhat, zkSync Era would need to implement its own version of the @ethereumjs/vm EVM implementation. Further investigation and implementation may be needed to determine the feasibility and impact of this feature on the zkSync Era network.
